### PR TITLE
Add faster raycast algorithm using DDA

### PR DESCRIPTION
**Pull request by nhatao**
This PR adds a faster raycast algorithm using DDA(Digital Differential Analyzer). 
DDA is a line drawing algorithm, and can enumerate all voxels that a line passes through.
The details of the DDA are described in [this paper](http://www.cse.yorku.ca/~amana/research/grid.pdf).
The name of variables are based on this paper.

The result of performance_raycast on my laptop is shown below.
The new algorithm is much faster than the current algorithm unless map is updated frequently.

> ## Chunk size: 1
> - Generate kdtree: 0.914859 sec
> - Collisions: 3722/7644
> - mcl_3dl::RaycastUsingKDTree Init: 3.8e-08 sec
> - mcl_3dl::RaycastUsingKDTree: 0.301234 sec
> 
> ## Chunk size: 2
> - Generate kdtree: 0.46443 sec
> - Collisions: 3722/7644
> - mcl_3dl::RaycastUsingKDTree Init: 4.2e-08 sec
> - mcl_3dl::RaycastUsingKDTree: 0.314594 sec
> 
> ## Chunk size: 5
> - Generate kdtree: 0.678015 sec
> - Collisions: 3722/7644
> - mcl_3dl::RaycastUsingKDTree Init: 4.1e-08 sec
> - mcl_3dl::RaycastUsingKDTree: 1.28824 sec
> 
> ## Chunk size: 10
> - Generate kdtree: 0.492192 sec
> - Collisions: 3722/7644
> - mcl_3dl::RaycastUsingKDTree Init: 3.7e-08 sec
> - mcl_3dl::RaycastUsingKDTree: 1.41755 sec
> 
> ## Chunk size: 20
> - Generate kdtree: 0.524135 sec
> - Collisions: 3722/7644
> - mcl_3dl::RaycastUsingKDTree Init: 5.7e-08 sec
> - mcl_3dl::RaycastUsingKDTree: 1.5855 sec
> 
> ## Chunk size: 50
> - Generate kdtree: 0.544604 sec
> - Collisions: 3722/7644
> - mcl_3dl::RaycastUsingKDTree Init: 3.8e-08 sec
> - mcl_3dl::RaycastUsingKDTree: 1.71085 sec
> 
> ## Chunk size: 100
> - Generate kdtree: 0.541325 sec
> - Collisions: 3722/7644
> - mcl_3dl::RaycastUsingKDTree Init: 3.8e-08 sec
> - mcl_3dl::RaycastUsingKDTree: 1.81081 sec
> 
> ## DDA grid size: 0.1
> - Generate kdtree: 0.469088 sec
> - Collisions: 3731/7644
> - mcl_3dl::RaycastUsingDDA Init: 0.813076 sec
> - mcl_3dl::RaycastUsingDDA: 0.0437835 sec
> 
> ## DDA grid size: 0.2
> - Generate kdtree: 0.473022 sec
> - Collisions: 3731/7644
> - mcl_3dl::RaycastUsingDDA Init: 0.175539 sec
> - mcl_3dl::RaycastUsingDDA: 0.0122947 sec
> 
> ## DDA grid size: 0.3
> - Generate kdtree: 0.459906 sec
> - Collisions: 3731/7644
> - mcl_3dl::RaycastUsingDDA Init: 0.110722 sec
> - mcl_3dl::RaycastUsingDDA: 0.00758433 sec
> 
> ## DDA grid size: 0.4
> - Generate kdtree: 0.460951 sec
> - Collisions: 3731/7644
> - mcl_3dl::RaycastUsingDDA Init: 0.0678405 sec
> - mcl_3dl::RaycastUsingDDA: 0.00503198 sec
> 
> ## DDA grid size: 0.5
> - Generate kdtree: 0.458588 sec
> - Collisions: 3731/7644
> - mcl_3dl::RaycastUsingDDA Init: 0.0681976 sec
> - mcl_3dl::RaycastUsingDDA: 0.00416095 sec

In addition, I modified the likelihood calculation algorithm of LidarMeasurementModelBeam.
It gives a penalty both when a range scan distance is shorter or longer than expected.
By default, the previous likelihood calculation algorithm (gives a penalty only when a range scan distance is shorter than expected) is used. (`add_penalty_short_only_mode : true`)

### CODE REVIEW

**Comment by at-wat** [comment]
## [[#859](https://travis-ci.org/at-wat/mcl_3dl/builds/735927797)] FAILED on noetic

Tested on Alpine ROS

**Comment by at-wat** [comment]
## [[#860](https://travis-ci.org/at-wat/mcl_3dl/builds/735932588)] PASSED on noetic

Tested on Alpine ROS

**Comment by nhatao** [comment]
@at-wat PTAL.

File: include/mcl_3dl/raycasts/raycast_using_dda.h [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+  {
+    if (!kdtree_->getInputCloud())
+    {
+      return;
+    }
+    if ((kdtree_->getInputCloud()->header.stamp == point_cloud_header_.stamp) && !points_.empty())
+    {
+      return;
+    }
+    const typename pcl::PointCloud<POINT_TYPE>::ConstPtr& input = kdtree_->getInputCloud();
```
Comment (at-wat) [inline include/mcl_3dl/raycasts/raycast_using_dda.h:176]: Converting whole map into gridmap requires large memory.
e.g. size of a largest map I have is 468x308x29m. It requires 2,339 * 1,540 * 147 * (1 (point_exists_) + 8 (points_)) = 530M * 9 = 4.7GB for 0.2m grid.
(pcd file size is 28MB)

This was why I didn't do it in the Kd-tree based implementation.

File: src/lidar_measurement_model_beam.cpp [hunk 2]
```diff
@@ -91,6 +96,46 @@ void LidarMeasurementModelBeam::loadConfig(
...
+
+  pnh.param("add_penalty_short_only_mode", add_penalty_short_only_mode_, true);
+  bool use_raycast_using_dda;
+  pnh.param("use_raycast_using_dda", use_raycast_using_dda, false);
+  if (use_raycast_using_dda)
+  {
+    double ray_angle_half;
+    pnh.param("ray_angle_half", ray_angle_half, 0.25 * M_PI / 180.0);
+    double dda_grid_size;
+    pnh.param("dda_grid_size", dda_grid_size, 0.2);
```
Comment (at-wat) [inline src/lidar_measurement_model_beam.cpp:108]: I think `dda_grid_size` should be greater than or equal to `max(map_grid_x_, map_grid_y_, map_grid_z_)`. Better to have a validation.

File: src/mcl_3dl.cpp [hunk 3]
```diff
@@ -135,6 +135,7 @@ class MCL3dlNode
     }
 
     pc_map_.reset(new pcl::PointCloud<PointType>);
+    pcl_conversions::toPCL(ros::Time::now(), pc_map_->header.stamp);
```
Comment (at-wat) [inline src/mcl_3dl.cpp:138]: Why it isn't
```suggestion
    pcl_conversions::toPCL(msg->header.stamp, pc_map_->header.stamp);
```
?

File: include/mcl_3dl/raycasts/raycast_using_dda.h (hunk 1 above)
Comment (at-wat) [reply include/mcl_3dl/raycasts/raycast_using_dda.h:176]: As an experimental feature, it's ok for now.

File: src/lidar_measurement_model_beam.cpp (hunk 2 above)
Comment (nhatao) [reply src/lidar_measurement_model_beam.cpp:108]: Done

File: src/mcl_3dl.cpp (hunk 3 above)
Comment (nhatao) [reply src/mcl_3dl.cpp]: Fixed to use msg->header.stamp when it is valid.

**Comment by at-wat** [comment]
## [[#861](https://travis-ci.org/at-wat/mcl_3dl/builds/735955700)] PASSED on noetic

Tested on Alpine ROS

File: include/mcl_3dl/raycasts/raycast_using_dda.h (hunk 1 above)
Comment (nhatao) [reply include/mcl_3dl/raycasts/raycast_using_dda.h:176]: The size of `points_` depends on the number of points of `pc_map_`(after downsampled), so additional memory consumption will become 530MB + maybe less than 10MB. 
To use this algorithm for larger map, it is needed to implement another method to reduce memory sizes.

**Review by at-wat — APPROVED** [review]
LGTM

File: src/mcl_3dl.cpp [hunk 4]
```diff
@@ -135,6 +135,8 @@ class MCL3dlNode
     }
 
     pc_map_.reset(new pcl::PointCloud<PointType>);
+    const ros::Time map_stamp = (msg->header.stamp == ros::Time()) ? ros::Time::now() : msg->header.stamp;
```
Comment (at-wat) [inline src/mcl_3dl.cpp:138]:
```suggestion
    const ros::Time map_stamp = msg->header.stamp.isValid() ? msg->header.stamp : ros::Time::now();
```
would be better.

**Comment by at-wat** [comment]
## [[#862](https://travis-ci.org/at-wat/mcl_3dl/builds/735962137)] PASSED on noetic

Tested on Alpine ROS

**Pull request merged**